### ISSUE TITLE
Build Failing with Error Message undeclared function call to 'ENGINE_load_builtin_engines'

### ISSUE THREAD

**Issue by jilgolfo** [janestreet/async_ssl#36]
Hi, on a fresh switch with ocaml 4.14.0 when installing async_ssl, I get the following error message:

```
[ERROR] The compilation of async_ssl.v0.15.0 failed at "dune build -p async_ssl -j 7".

#=== ERROR while compiling async_ssl.v0.15.0 ==================================#
# context     2.1.4 | macos/arm64 | ocaml-base-compiler.4.14.0 | https://opam.ocaml.org#331309a7
# path        ~/.opam/test-async-2/.opam-switch/build/async_ssl.v0.15.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p async_ssl -j 7
# exit-code   1
# env-file    ~/.opam/log/async_ssl-51863-2f3d8f.env
# output-file ~/.opam/log/async_ssl-51863-2f3d8f.out
### output ###
# 16 |  (c_names ffi_generated_stubs) (flags :standard -w -9-27-32-34)
# [...]
# (cd _build/default/src && /usr/bin/cc -O2 -fno-strict-aliasing -fwrapv -pthread -D_FILE_OFFSET_BITS=64 -O2 -fno-strict-aliasing -fwrapv -pthread -w -I/opt/homebrew/Cellar/openssl@3/3.1.0/include -g -I /Users/myname/.opam/test-async-2/lib/ocaml -I /Users/myname/.opam/test-async-2/lib/async -I /Users/myname/.opam/test-async-2/lib/async/async_command -I /Users/myname/.opam/test-a[...]
# ffi_generated_stubs.c:80:4: error: call to undeclared function 'ENGINE_load_builtin_engines'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
#    ENGINE_load_builtin_engines();
#    ^
# ffi_generated_stubs.c:85:4: error: call to undeclared function 'ENGINE_unregister_RAND'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
#    ENGINE_unregister_RAND();
#    ^
# ffi_generated_stubs.c:90:4: error: call to undeclared function 'ENGINE_register_all_complete'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
#    ENGINE_register_all_complete();
#    ^
# 3 errors generated.
```

I am on a Mac M2, MacOS Ventura 13.3.1.

Any ideas what to do? Thank you!

**Comment by anuragsoni** [comment]
I'm also seeing this error on intel macOS. This might be related to `homebrew` updating openssl to version 3.1 as I was able to install and use `async_ssl` without any issue with openssl 3.0.x.

**Comment by anuragsoni** [comment]
I spent some more time looking at openssl, and I was able to reproduce this with an older openssl version on macOS as well. I am no longer sure why/how async_ssl compiled on macOS when I installed it earlier this year :( 

That said, I was able to get further in the build by including `openssl/engine.h` in the c files generated by the `stubgen` for async_ssl.

```diff
diff --git a/stubgen/ffi_stubgen.ml b/stubgen/ffi_stubgen.ml
index a8739ed..c8db2b6 100644
--- a/stubgen/ffi_stubgen.ml
+++ b/stubgen/ffi_stubgen.ml
@@ -7,6 +7,7 @@ let prologue =
    #include <openssl/ssl.h>\n\
    #include <openssl/err.h>\n\
    #include <openssl/conf.h>\n\n\
+   #include <openssl/engine.h>\n\n\
    #include \"../bindings/openssl_helpers.h\"\n"
 ;;
 
```

This leads to another compilation error:

```
File "src/dune", line 16, characters 10-29:
16 |  (c_names ffi_generated_stubs) (flags :standard -w -9-27-32-34)
               ^^^^^^^^^^^^^^^^^^^
ffi_generated_stubs.c:87:27: error: too few arguments to function call, single argument 'e' was not specified
   ENGINE_unregister_RAND();
   ~~~~~~~~~~~~~~~~~~~~~~ ^
/opt/homebrew/Cellar/openssl@3/3.1.0/include/openssl/engine.h:391:28: note: 'ENGINE_unregister_RAND' declared here
OSSL_DEPRECATEDIN_3_0 void ENGINE_unregister_RAND(ENGINE *e);
```

This error also makes sense as `ENGINE_unregister_RAND ` indeed accepts an ENGINE* as an argument (https://github.com/openssl/openssl/blob/3868807d2fe5a72aa897ce5f7f7ba7e9cc3c09cb/include/openssl/engine.h#L391) and it has done so since the 1.0 days.

**Comment by anuragsoni** [comment]
```diff
commit cafffb990fe5f39a09bdb053481e777ed8427c1f (HEAD -> refs/heads/v0.15)
Author: Anurag Soni <anurag@sonianurag.com>
Date:   Mon May 8 21:09:13 2023 -0400

    Write helper to unregister rand from all engines

diff --git a/bindings/ffi_bindings.ml b/bindings/ffi_bindings.ml
index 9da4c9f..67f3779 100644
--- a/bindings/ffi_bindings.ml
+++ b/bindings/ffi_bindings.ml
@@ -323,7 +323,7 @@ module Bindings (F : Cstubs.FOREIGN) = struct
     ;;
 
     let unregister_RAND =
-      foreign "ENGINE_unregister_RAND" Ctypes.(void @-> returning void)
+      foreign "async_ssl__unregister_rand" Ctypes.(void @-> returning void)
     ;;
 
     let register_all_complete =
diff --git a/bindings/openssl_helpers.c b/bindings/openssl_helpers.c
index dfa899d..0a9d962 100644
--- a/bindings/openssl_helpers.c
+++ b/bindings/openssl_helpers.c
@@ -1,6 +1,7 @@
 #include <string.h>
 #include <openssl/x509v3.h>
 #include <openssl/ssl.h>
+#include <openssl/engine.h>
 #include "openssl_helpers.h"
 
 /* Return a NULL-terminated array of strings representing the subjectAltNames in [cert].
@@ -148,3 +149,10 @@ void async_ssl__free_pem_peer_certificate_chain(char *certs) {
     free(certs);
     return;
 }
+
+
+void async_ssl__unregister_rand() {
+    ENGINE *e;
+    for (e = ENGINE_get_first(); e; e = ENGINE_get_next(e))
+        ENGINE_unregister_RAND(e);
+}
diff --git a/bindings/openssl_helpers.h b/bindings/openssl_helpers.h
index 514f0c8..e0e18d1 100644
--- a/bindings/openssl_helpers.h
+++ b/bindings/openssl_helpers.h
@@ -2,3 +2,4 @@ char **async_ssl__subject_alt_names(const X509 *cert);
 void async_ssl__free_subject_alt_names(char **results);
 char *async_ssl__pem_peer_certificate_chain(const SSL *con);
 void async_ssl__free_pem_peer_certificate_chain(char *certs);
+void async_ssl__unregister_rand();
```

With this patch applied, I'm able to compile async_ssl on macOS and validate that I can still use async_ssl to connect to remote servers that use https.

**Comment by anuragsoni** [comment]
> We'll include the fix in v0.16.1 (since v0.16.0 is already on its way to the main opam repo).

Thank you!

**Comment by dsull800** [comment]
I made the changes that @anuragsoni suggested but now I received an error 'Unbound module Time_float' whenever I ran `opam install .`, my solution was to change `Time_float` to `Time_unix` in bindings/src/ssl.ml.

**Comment by dkalinichenko-js** [comment]
Fixed in v0.16.1.

**Comment by anuragsoni** [comment]
> Fixed in v0.16.1.

Thank you!
